libs/infinity_emb/infinity_emb/transformer/audio/torch.py

```diff
@@ -0,0 +1,151 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2023-now michaelfeil
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Iterable, Optional, Union
+
+from infinity_emb._optional_imports import CHECK_TORCH, CHECK_TRANSFORMERS
+from infinity_emb.args import EngineArgs
+from infinity_emb.transformer.abstract import BaseClapAudioModel
+from infinity_emb.transformer.quantization.interface import quant_embedding_decorator
+
+if TYPE_CHECKING:
+    import numpy as np
+    from torch import Tensor
+
+if CHECK_TORCH.is_available:
+    import torch
+if CHECK_TRANSFORMERS.is_available:
+    from transformers import AutoModel, AutoProcessor  # type: ignore
+
+
+class ClapLikeModel(BaseClapAudioModel):
+    """CrossEncoder with .encode_core() and no microbatching"""
+
+    def __init__(self, *, engine_args: EngineArgs):
+        CHECK_TORCH.mark_required()
+        CHECK_TRANSFORMERS.mark_required()
+        self.model = AutoModel.from_pretrained(
+            engine_args.model_name_or_path,
+            revision=engine_args.revision,
+            trust_remote_code=engine_args.trust_remote_code,
+            # attn_implementation="eager" if engine_args.bettertransformer else None,
+        )
+
+        # self.model = to_bettertransformer(
+        #     self.model,
+        #     engine_args,
+        #     logger,
+        # )
+        self.processor = AutoProcessor.from_pretrained(
+            engine_args.model_name_or_path,
+            revision=engine_args.revision,
+            trust_remote_code=engine_args.trust_remote_code,
+        )
+        self.engine_args = engine_args
+
+        if engine_args.compile:
+            self.model.vision_model = torch.compile(
+                self.model.vision_model, dynamic=True
+            )
+            self.model.text_model = torch.compile(self.model.text_model, dynamic=True)
+
+        assert hasattr(
+            self.model, "get_text_features"
+        ), f"AutoModel of {engine_args.model_name_or_path} does not have get_text_features method"
+        assert hasattr(
+            self.model, "get_audio_features"
+        ), f"AutoModel of {engine_args.model_name_or_path} does not have get_audio_features method"
+        self.max_length = None
+        if hasattr(self.model.config, "max_length"):
+            self.max_length = self.model.config.max_length
+        elif hasattr(self.model.config, "max_position_embeddings"):
+            self.max_length = self.model.config.max_position_embeddings
+        elif hasattr(self.model.config, "text_config") and hasattr(
+            self.model.config.text_config, "max_length"
+        ):
+            self.max_length = self.model.config.text_config.max_length
```

Comment:
Maybe set it to `self.max_length = max(0, self.max_length - 3)` to avoid the tokenization issue you were commenting about.

Reply:
With the current version of using the max_length it seems to work.
```
        loaded_engine.warmup(
            batch_size=engine_args.batch_size, n_tokens=loaded_engine.max_length
        )
        emb_per_sec, _, log_msg = loaded_engine.warmup(
            batch_size=engine_args.batch_size, n_tokens=loaded_engine.max_length
        )
```

Previously the n_tokens was set to a fixed value of 512 which caused the problems.

Reply:
Regarding the tolerance, seems I was wrong regarding multiple passes differ. I checked thats not the case with this code:
```
    model_name = "laion/clap-htsat-unfused"
    model = ClapModel.from_pretrained(model_name)
    processor = ClapProcessor.from_pretrained(model_name)

    max_length = model.config.text_config.max_length
    inputs_clap = processor(
        text=["a sound of a cat"],
        audios=[data],
        return_tensors="pt",
        padding="max_length",
        sampling_rate=48000,
        max_length=max_length,
    )

    outputs = model(**inputs_clap)
    outputs2 = model(**inputs_clap)

    torch.testing.assert_close(
        outputs.audio_embeds[0], outputs2.audio_embeds[0], check_dtype=False, rtol=0, atol=1e-4
    )
```
Here 1e-4 passes. 

Need to dig deeper. Can I assume that `AutoProcessor.from_pretrained` does the same as 'ClapProcessor.from_pretrained(model_name)'?

Reply:
RE: Can I assume that AutoProcessor.from_pretrained does the same as 'ClapProcessor.from_pretrained(model_name)'?


The right class for the processor is fetched based on the architecture. 
```json
"architectures": [ "ClapModel"]
```

https://huggingface.co/laion/clap-htsat-unfused/blob/main/config.json ClapModel should then be mapped to ClapProcessor in the https://github.com/huggingface/transformers/blob/9db963aeed419c8379c6d6425186fec0bfb86908/src/transformers/models/auto/modeling_auto.py#L61 config here.

Reply:
Ok, I think I got your point reg. the 512 fixed token length. We also need to guard the model in general, not only during the warmup phase. 

Whats your suggested path to achieve that?

Reply:
audio_or_text[:self.max_length] would limit it to 512 characters. Setting max_length in the tokenizer is the only thing to do here to preserve the special tokens at the beginning and end.

Reply:
I was assuming the `max_length` parameter would be bassed down to the tokenizer as kwargs? I was checking the Base Class of the ([RobertaTokenizerFast](https://huggingface.co/docs/transformers/v4.44.2/en/model_doc/roberta#transformers.RobertaTokenizerFast)) and I found the `max_length` [here](https://huggingface.co/docs/transformers/v4.44.2/en/internal/tokenization_utils#transformers.PreTrainedTokenizerBase) in the `__call__()` method. Is passing `truncation=True` required for this to limit the size? This results in a Not Implemented Error.

Might this be because  The [ClapFeatureExtractor](https://huggingface.co/docs/transformers/v4.44.2/en/model_doc/clap#transformers.ClapFeatureExtractor) and the tokenizer get the same parameters, but one expects True or False whereas the Extractor expects `fustion` or `rand_trunc`?